### PR TITLE
Ebpf xdp 5.0 v2.8

### PR DESCRIPTION
**Pull request by regit**
Rebase of #3904 on latest master.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- rebase on top of latest master

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/460
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/241

### CODE REVIEW

File: src/util-ebpf.c [hunk 1]
```diff
@@ -496,13 +496,18 @@ int EBPFSetupXDP(const char *iface, int fd, uint8_t flags)
     return 0;
 }
 
-static int EBPFCreateFlowForKey(struct flows_stats *flowstats, FlowKey *flow_key,
+/**
+ * Create a Flow in the table for a Flowkey
+ *
+ * \return false (or true if the Flow needs to be destroyed)
```
Comment (victorjulien) [inline src/util-ebpf.c]: it seems we always return false?

File: src/util-ebpf.c [hunk 2]
```diff
@@ -604,13 +614,14 @@ static int EBPFForEachFlowV4Table(ThreadVars *th_v, LiveDevice *dev, const char
     }
 
     uint64_t pkts_cnt = 0;
+    bool dead_flow = false;
     while (bpf_map_get_next_key(mapfd, &key, &next_key) == 0) {
         uint64_t bytes_cnt = 0;
         hash_cnt++;
-        if (pkts_cnt > 0) {
+        if (dead_flow > 0) {
```
Comment (victorjulien) [inline src/util-ebpf.c]: just
```c
if (dead_flow) {
```

File: src/util-ebpf.c [hunk 3]
```diff
@@ -604,13 +614,14 @@ static int EBPFForEachFlowV4Table(ThreadVars *th_v, LiveDevice *dev, const char
...
             EBPFDeleteKey(mapfd, &key);
         }
-        pkts_cnt = 0;
+        dead_flow = false;
```
Comment (victorjulien) [inline src/util-ebpf.c]: can be moved into the if (dead_flow) { condition

File: src/util-ebpf.c (hunk 1 above)
Comment (regit) [reply src/util-ebpf.c]: Yes, but function from same family are returning True if Flow needs to be destroyed.

Comment (victorjulien) [reply src/util-ebpf.c]: The comment caused my confusion. What about: '\return false (this create function never returns true)'

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 4]
```diff
@@ -350,11 +375,135 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+they can also be accessed by external tools. In Suricata bypass case, this can be
+used to keep active bypassed flow tables so Suricata is not hit by previsouly bypassed flows when
+restarting. In the socket filter case, this can be used to maintain a map from tools outside
+of Suricata.
+
+To used pinned maps, you first have to mount the `bpf` pseudo filesystem ::
+
+  sudo mount -t bpf none /sys/fs/bpf
+
+You can also add to you `/etc/fstab/` ::
```
Comment (victorjulien) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: ```/etc/fstab``` (drop trailing /)

**Comment by victorjulien** [comment]
Playing with this I've run into this error/warning output that needs a bit of cleanup:
```
[4201] 4/6/2019 -- 12:54:17 - (runmode-af-packet.c:275) <Info> (ParseAFPConfig) -- AF_PACKET IPS mode activated ens5np0->ens5np1
[4201] 4/6/2019 -- 12:54:17 - (runmode-af-packet.c:479) <Info> (ParseAFPConfig) -- af-packet will use '/home/victor/dev/suricata/ebpf/xdp_filter.bpf' as XDP filter file
libbpf: failed to create map (name: 'flow_table_v4'): Invalid argument
libbpf: failed to load object '/home/victor/dev/suricata/ebpf/xdp_filter.bpf'
[4201] 4/6/2019 -- 12:54:17 - (util-ebpf.c:394) <Error> (EBPFLoadFile) -- [ERRCODE: SC_ERR_MEM_ALLOC(1)] - Permission issue when loading eBPF object: Unknown error -1 (-1)
[4201] 4/6/2019 -- 12:54:17 - (runmode-af-packet.c:539) <Warning> (ParseAFPConfig) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Error when loading XDP filter file
[4201] 4/6/2019 -- 12:54:17 - (util-runmodes.c:297) <Info> (RunModeSetLiveCaptureWorkersForDevice) -- Going to use 12 thread(s)
[4425] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4425] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4428] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4428] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4435] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4435] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4440] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4440] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4445] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4445] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4452] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4452] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4457] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4457] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4464] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4464] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4475] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4475] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4481] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4481] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4485] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4485] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'
[4497] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2608) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v4'
[4497] 4/6/2019 -- 12:54:17 - (source-af-packet.c:2612) <Error> (ReceiveAFPThreadInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Can't find eBPF map fd for 'flow_table_v6'

```

File: src/util-ebpf.c (hunk 1 above)
Comment (regit) [reply src/util-ebpf.c]: That's better indeed

**Comment by victorjulien** [comment]
Ok as discussed offline, there is a tricky situation with the pinned-maps and modifications to the filter. In short, when the filter is loaded and the pinned maps are created, the filter stays active when Suricata stops, but also when Suricata starts again. For modifications to the filter to go into effect, the pinned maps (sitting in ```/sys/fs/bpf/suricata*```) will have to be removed. This *is* mentioned in the docs, but it's still far too easy to loose time on.

The simplest workaround is probably to detect if the loading of maps was successful and in that case give a message that the already loaded filter is being used.

I wonder if it could be made smarter though, perhaps with the help of suricatactl. One thought is to put the actually loaded bpf binary in ```/var/lib/suricata/bpf/``` and load it from there. Then use some new suricatactl command to update the filter file and remove the maps, or perhaps leave a message in the bpf dir for the init script to remove the maps on restart. These are incomplete thoughts but I think for optimal usability it would be good to improve the end-user workflow significantly.

**Comment by victorjulien** [comment]
Another update: I could not get the nfp support to work in either driver or hw mode on Ubuntu 18.04 with the default 4.15 kernel. With their 5.0.0-15-generic I had better luck. This can be installed using ```apt install linux-generic-hwe-18.04-edge```

**Comment by victorjulien** [comment]
This should probably be a counter? Seeing very many of them.
```
[11135] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11127] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11133] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11128] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11126] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11135] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11125] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11134] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11127] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11133] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11128] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11135] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11134] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11125] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11126] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11128] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11125] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11133] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11134] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11127] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)
[11126] 4/6/2019 -- 17:24:17 - (source-af-packet.c:2325) <Error> (AFPInsertHalfFlow) -- [ERRCODE: SC_ERR_BPF(127)] - Can't update eBPF map: Resource temporarily unavailable (11)

```

**Comment by victorjulien** [comment]
What does this error mean? Is the table full? It's trivial to fill it currently, takes a second or so using trex.

**Comment by regit** [comment]
> What does this error mean? Is the table full? It's trivial to fill it currently, takes a second or so using trex.

I think it is because it is full but it was not supposed to be EGAIN error. I'm checking what is happening here.

You can increase the size of the flow table in the maps declaration in xdp_filter.c

**Comment by victorjulien** [comment]
How can I determine how high to set this? I assume there are limits, but can I know what to set it to w/o a trial and error loop?

**Comment by victorjulien** [comment]
A bit of trial and error: 1.5M map size for ipv4 works, 2M map size fails on nfp.

Still able to get it to run out.

**Comment by victorjulien** [comment]
This pkts vs bytes different looks a bit suspicious:
```
flow_bypassed.closed                          | Total                     | 4362995
flow_bypassed.pkts                            | Total                     | 38356617952123
flow_bypassed.bytes                           | Total                     | 7442671369
```

**Comment by victorjulien** [comment]
Left it running for a bit:
```
flow_bypassed.closed                          | Total                     | 23980491
flow_bypassed.pkts                            | Total                     | 189260138580011
flow_bypassed.bytes                           | Total                     | 24579945826
```

**Comment by victorjulien** [comment]
Left it running overnight:
```
flow_bypassed.closed                          | Total                     | 237957573
flow_bypassed.pkts                            | Total                     | 2169743913481361
flow_bypassed.bytes                           | Total                     | 278766329486
```
```
# ethtool -S ens5np0|grep bpf
     bpf_pass_pkts: 51616379775
     bpf_pass_bytes: 10728847894276
     bpf_app1_pkts: 7438754627
     bpf_app1_bytes: 1384341064987
     bpf_app2_pkts: 0
     bpf_app2_bytes: 0
     bpf_app3_pkts: 0
     bpf_app3_bytes: 0
```
So the card tells us 'app1' passed 7.4 billion packets. Should that relate to flow_bypassed.pkts?

**Comment by regit** [comment]
The pkts counter seems to be off. It definitely should be bigger than bytes.

**Comment by regit** [comment]
OK fixed the counter system, there was a big mistake.

**Comment by regit** [comment]
Regarding the EAGAIN (error 11) it is not linked to the size of the hash. We got ENOSPC if the hash table is full.

**Comment by victorjulien** [comment]
It seems to have some kind of correlation. When I increase the mapsize I see fewer of these EAGAIN errors, and it seems to take longer before they appear. Maybe there is something nfp specific going on.

**Comment by regit** [comment]
> It seems to have some kind of correlation. When I increase the mapsize I see fewer of these EAGAIN errors, and it seems to take longer before they appear. Maybe there is something nfp specific going on.

Interesting it may worth to ask people from netronome.

**Comment by regit** [comment]
Closing and opening with fixes.

**Pull request closed without merging**